### DISCUSSION TITLE
Matching of ip addresses to ip ranges not as expected

### DISCUSSION THREAD

**Discussion by funked00** [netbox-community/netbox#15639] [Help Wanted!] [answered]
I have found the recommendation in several places to save an ip address with the subnet mask that it also has on the interface. However, automatic matching into an IP range does not work in a meaningful way.
Example:
subnet 172.19.50.0/23
ip range 172.19.50.1/32 - 172.19.50.255/32
ip range 172.19.51.0/32 - 172.19.51.254/32

ip address 172.19.50.100/23 does not appear in the ip range.
ip address 172.19.50.100/32 appears in the ip range.

Do you have any ideas or other experiences about this?

Best regards
Daniel

**Comment by candlerb** [comment, marked as answer]
Yeah, it's a bit of weirdness with Netbox that IP addresses are not just an IP address, but IP address + prefix length. No other IPAM system I've ever seen does that.

I think you're expected to do the same with ranges too. That is, your ranges should be

ip range 172.19.50.1/23 - 172.19.50.255/23
ip range 172.19.51.0/23 - 172.19.51.254/23

**Reply by jeremystretch** [reply]
> Yeah, it's a bit of weirdness with Netbox that IP addresses are not just an IP address, but IP address + prefix length. No other IPAM system I've ever seen does that.

[Why does NetBox Require Every IP to Have a Mask?](https://github.com/netbox-community/netbox/wiki/Frequently-Asked-Questions#why-does-netbox-require-every-ip-to-have-a-mask)

**Reply by funked00** [reply]
Unfortunately, this only works if I only define one ip range. If I create both ranges with /23, they are probably congruent according to the netbox logic and therefore the IP address can no longer be clearly assigned.

**Reply by candlerb** [reply]
> Unfortunately, this only works if I only define one ip range. If I create both ranges with /23, they are probably congruent according to the netbox logic and therefore the IP address can no longer be clearly assigned.

Sorry, I don't understand what you're trying to say. If you create the two IP ranges that I specified, what problem do you observe?

An IP range of "172.19.50.1/23 - 172.19.50.255/23" is actually just a block of 255 addresses from 172.19.50.1 to 172.19.50.255 inclusive, notated with Netbox's weird idea that every individual IP address must also be associated with a prefix length. But whatever value you give to this prefix length makes no difference to the range of address covered. It won't overlap with the range "172.19.51.0/23 - 172.19.51.254/23" (since both start and end address of the latter range are greater than the former)

In a range, the masks given to the start and end address [must match](https://github.com/netbox-community/netbox/blob/v3.7.5/netbox/ipam/models/ip.py#L567-L571).  Also, ranges [must not overlap](https://github.com/netbox-community/netbox/blob/v3.7.5/netbox/ipam/models/ip.py#L581-L592) within the same VRF, but no consideration is given to the mask when evaluating this; only the addresses themselves define the start and end of the range.

**Reply by funked00** [reply]
You're absolutely right, I can't recreate what I saw, I guess I was confused. Now it works with /23 on the ranges. Thanks a lot for your help!
